File: convert-to-pdf.py
#!/usr/bin/env python3
"""
Convertidor de Markdown a PDF
Requiere: pip install markdown2 pdfkit wkhtmltopdf
"""

import os
import sys
import subprocess
from pathlib import Path

def convert_md_to_pdf():
    script_dir = Path(__file__).parent
    md_file = script_dir / "PRESUPUESTO.md"
    pdf_file = script_dir / "PRESUPUESTO.pdf"
    
    if not md_file.exists():
        print(f"❌ Archivo no encontrado: {md_file}")
        return False
    
    try:
        # Intenta con pandoc primero (mejor calidad)
        try:
            cmd = [
                'pandoc',
                str(md_file),
                '-o', str(pdf_file),
                '--from=markdown',
                '--to=pdf',
                '-V', 'margin-top=20mm',
                '-V', 'margin-bottom=20mm',
                '-V', 'margin-left=15mm',
                '-V', 'margin-right=15mm',
                '--highlight-style=pygments'
            ]
            subprocess.run(cmd, check=True, capture_output=True)
            print(f"✅ PDF generado con pandoc: {pdf_file}")
            return True
        except (subprocess.CalledProcessError, FileNotFoundError):
            pass
        
        # Alternativa con markdown2 + pdfkit + wkhtmltopdf
        try:
            import markdown2
            import pdfkit
            
            with open(md_file, 'r', encoding='utf-8') as f:
                md_content = f.read()
            
            html_content = markdown2.markdown(md_content)
            
            html_with_style = f"""
            <!DOCTYPE html>
            <html>
            <head>
                <meta charset="UTF-8">
                <style>
                    body {{
                        font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
                        line-height: 1.6;
                        color: #333;
                        margin: 0;
                        padding: 20px;
                    }}
                    h1, h2, h3, h4, h5, h6 {{
                        color: #0066cc;
                        margin-top: 25px;
                        margin-bottom: 15px;
                    }}
                    h1 {{ font-size: 28px; border-bottom: 3px solid #0066cc; padding-bottom: 10px; }}
                    h2 {{ font-size: 22px; }}
                    h3 {{ font-size: 18px; }}
                    table {{
                        width: 100%;
                        border-collapse: collapse;
                        margin: 20px 0;
                    }}
                    th, td {{
                        border: 1px solid #ddd;
                        padding: 12px;
                        text-align: left;
                    }}
                    th {{
                        background-color: #f0f0f0;
                        font-weight: bold;
                        color: #0066cc;
                    }}
                    tr:nth-child(even) {{ background-color: #f9f9f9; }}
                    code {{
                        background-color: #f4f4f4;
                        padding: 2px 6px;
                        border-radius: 3px;
                        font-family: 'Courier New', monospace;
                    }}
                    blockquote {{
                        border-left: 4px solid #0066cc;
                        padding-left: 20px;
                        color: #666;
                    }}
                </style>
            </head>
            <body>
                {html_content}
            </body>
            </html>
            """
            
            options = {
                'page-size': 'A4',
                'margin-top': '15mm',
                'margin-right': '15mm',
                'margin-bottom': '15mm',
                'margin-left': '15mm',
                'encoding': 'UTF-8',
            }
            
            pdfkit.from_string(html_with_style, str(pdf_file), options=options)
            print(f"✅ PDF generado con pdfkit: {pdf_file}")
            return True
        except ImportError:
            pass
        
        # Si nada funcionó, mostrar error
        print("❌ No se encontraron herramientas de conversión.")
        print("Instala una de estas opciones:")
        print("  - pandoc: https://pandoc.org/installing.html")
        print("  - pdfkit + wkhtmltopdf: pip install pdfkit && choco install wkhtmltopdf")
        return False
        
    except Exception as e:
        print(f"❌ Error: {e}")
        return False

if __name__ == "__main__":
    success = convert_md_to_pdf()
    sys.exit(0 if success else 1)

File: convert-to-pdf.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Intentar con marked y html-pdf si están disponibles
try {
  const marked = require('marked');
  const htmlPdf = require('html-pdf');

  const mdFile = path.join(__dirname, 'PRESUPUESTO.md');
  const pdfFile = path.join(__dirname, 'PRESUPUESTO.pdf');

  // Leer el archivo markdown
  const markdownContent = fs.readFileSync(mdFile, 'utf-8');

  // Convertir markdown a HTML
  const htmlContent = marked.parse(markdownContent);

  // Envolver en HTML completo con CSS
  const fullHtml = `
    <!DOCTYPE html>
    <html>
    <head>
      <meta charset="UTF-8">
      <style>
        body {
          font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
          line-height: 1.6;
          color: #333;
          margin: 0;
          padding: 20px;
          background: white;
        }
        h1, h2, h3, h4, h5, h6 {
          color: #0066cc;
          margin-top: 30px;
          margin-bottom: 15px;
        }
        h1 { font-size: 28px; border-bottom: 3px solid #0066cc; padding-bottom: 10px; }
        h2 { font-size: 22px; margin-top: 25px; }
        h3 { font-size: 18px; }
        h4 { font-size: 16px; }
        p { margin: 10px 0; text-align: justify; }
        ul, ol { margin: 15px 0; padding-left: 30px; }
        li { margin: 8px 0; }
        table {
          width: 100%;
          border-collapse: collapse;
          margin: 20px 0;
          page-break-inside: avoid;
        }
        th, td {
          border: 1px solid #ddd;
          padding: 12px;
          text-align: left;
        }
        th {
          background-color: #f0f0f0;
          font-weight: bold;
          color: #0066cc;
        }
        tr:nth-child(even) { background-color: #f9f9f9; }
        code, tt {
          background-color: #f4f4f4;
          padding: 2px 6px;
          border-radius: 3px;
          font-family: 'Courier New', monospace;
        }
        blockquote {
          border-left: 4px solid #0066cc;
          margin: 15px 0;
          padding-left: 20px;
          color: #666;
          font-style: italic;
        }
        hr {
          border: none;
          border-top: 2px solid #0066cc;
          margin: 30px 0;
          page-break-after: avoid;
        }
        strong { color: #0066cc; font-weight: bold; }
        em { font-style: italic; }
        .page-break { page-break-after: always; }
      </style>
    </head>
    <body>
      ${htmlContent}
    </body>
    </html>
  `;

  const options = {
    format: 'A4',
    margin: {
      top: '15mm',
      right: '15mm',
      bottom: '15mm',
      left: '15mm'
    },
    border: {
      top: '0',
      right: '0',
      bottom: '0',
      left: '0'
    }
  };

  htmlPdf.create(fullHtml, options).toFile(pdfFile, (err, res) => {
    if (err) {
      console.error('Error generando PDF:', err);
      process.exit(1);
    }
    console.log(`✅ PDF generado exitosamente: ${pdfFile}`);
    process.exit(0);
  });

} catch (error) {
  console.error('Dependencias faltantes. Instala:', error.message);
  console.log('\nIntentando instalación automática...');
  
  const { execSync } = require('child_process');
  try {
    execSync('npm install marked html-pdf', { stdio: 'inherit' });
    console.log('✅ Dependencias instaladas. Reinicia el script.');
  } catch (e) {
    console.error('No se pudo instalar automáticamente.');
    process.exit(1);
  }
}
